mpitest/test_fileref_mpi.py

```diff
@@ -0,0 +1,136 @@
+        for i in range(num_ins):
+            self.add_param("fin%d"%i, FileRef(name+'%d.in'%i))
+
+    def solve_nonlinear(self, params, unknowns, resids):
+        pass
+
+class FileRefTestCase(MPITestCase):
+    N_PROCS=4
+
+    def setUp(self):
+        self.startdir = os.getcwd()
+
+        if self.comm.rank == 0:
+            self.tmpdir = mkdtemp()
+        else:
+            self.tmpdir = None
+
+        if MPI:
+            #make sure everyone is using the same temp directory
+            self.tmpdir = self.comm.bcast(self.tmpdir, root=0)
+
+        os.chdir(self.tmpdir)
+
+    def tearDown(self):
+        if MPI:
+            # make sure we're done checking file contents before anyone deletes
+            # the tmp directory
+            self.comm.barrier()
+
+        os.chdir(self.startdir)
+        if self.comm.rank == 0:
+            try:
+                rmtree(self.tmpdir)
+            except OSError as e:
+                # If directory already deleted, keep going
+                if e.errno not in (errno.ENOENT, errno.EACCES, errno.EPERM):
+                    raise e
+
+    def test_file_diamond(self):
+        if MPI:
+            num = self.N_PROCS
+        else:
+            num = 1
+
+        prob = Problem(Group(), impl=impl)
+
+        src = prob.root.add("src", FileSrc('src'))
+        par = prob.root.add('par', ParallelGroup())
+        sink = prob.root.add("sink", FileSink('sink', num))
+
+        for i in range(num):
+            par.add("mid%d"%i, FileMid('mid%d'%i,'mid%d'%i))
+            prob.root.connect('src.fout', 'par.mid%d.fin'%i)
+            prob.root.connect('par.mid%d.fout'%i, 'sink.fin%d'%i)
+
+        prob.setup(check=False)
+        prob.run()
+
+        for i in range(num):
+            with sink.params['fin%d'%i].open('r') as f:
+                self.assertEqual(f.read(), "src\npar.mid%d\n"%i)
+
+    def test_file_diamond_same_names(self):
+        if MPI:
+            num = self.N_PROCS
+            directory = lambda comm: str(self.comm.rank)
+        else:
+            num = 1
+            directory = ''
+
+        prob = Problem(Group(), impl=impl)
+
+        src = prob.root.add("src", FileSrc('foo'))
+        par = prob.root.add('par', ParallelGroup())
+        sink = prob.root.add("sink", FileSink('sink', num))
+
+        for i in range(num):
+            # all FileMids will have output file with the same name, so
+            # framework needs to create rank specific directories for
+            # each rank to avoid collisions.
+            mid = par.add("mid%d"%i, FileMid('foo','foo'))
+            mid.directory = directory
+            mid.create_dirs = True
+            prob.root.connect('src.fout', 'par.mid%d.fin'%i)
+            prob.root.connect('par.mid%d.fout'%i, 'sink.fin%d'%i)
+
+        prob.setup(check=False)
+        prob.run()
+
+        for i in range(num):
+            with sink.params['fin%d'%i].open('r') as f:
+                self.assertEqual(f.read(), "src\npar.mid%d\n"%i)
```

Comment:
these tests are a bit obtuse as to what they are actually checking. Add a few comments to clarify